### DISCUSSION TITLE
Failed to upload images to third-party S3 service like Alibaba Cloud OSS

### DISCUSSION THREAD

**Discussion by KooriMoe** [outline/outline#3159] [Self Hosting]
I chose to use Alibaba Cloud OSS as S3 storage.
It can normally upload personal avatars and team pictures to Alibaba Cloud OSS but cannot upload files in documents.
In the browser, Outline will automatically merge the URL into the AWS S3 service link with ```amazonaws.com```, for example, ```inotea.s3.oss-cn-beijing.amazonaws.com```, but my S3 is ```inotea.s3.oss-cn-beijing.aliyuncs.com```
Is there a way to fix it?

### .env
```
AWS_ACCESS_KEY_ID=xxxxxxxxxxxxx
AWS_SECRET_ACCESS_KEY=xxxxxxxxxxxxx
AWS_REGION=oss-cn-beijing
AWS_S3_UPLOAD_BUCKET_URL=https://inotea.oss-cn-beijing.aliyuncs.com
AWS_S3_UPLOAD_BUCKET_NAME=inotea
AWS_S3_UPLOAD_MAX_SIZE=26214400
AWS_S3_FORCE_PATH_STYLE=false
AWS_S3_ACL=public
```

### Logs from Docker Container
```
yarn run v1.22.15
$ node ./build/server/index.js
{"label":"lifecycle","level":"info","message":"Note: Restricting process count to 1 due to use of collaborative service"}
{"label":"lifecycle","level":"info","message":"\nIs your team enjoying Outline? Consider supporting future development by sponsoring the project:\n\nhttps://github.com/sponsors/outline\n"}
{"label":"lifecycle","level":"info","message":"Note: Restricting process count to 1 due to use of collaborative service"}
{"label":"lifecycle","level":"info","message":"Starting collaboration service"}
{"label":"lifecycle","level":"info","message":"Starting websockets service"}
{"label":"lifecycle","level":"info","message":"Starting worker service"}
{"label":"lifecycle","level":"info","message":"Starting web service"}
{"level":"warn","message":"Enforced https was disabled with FORCE_HTTPS env variable"}
{"label":"lifecycle","level":"info","message":"Listening on http://localhost:3000"}
{"error":"Inaccessible host: `inotea.s3.oss-cn-beijing.amazonaws.com'. This service may not be available in the `oss-cn-beijing' region.","stack":"UnknownEndpoint: Inaccessible host: `inotea.s3.oss-cn-beijing.amazonaws.com'. This service may not be available in the `oss-cn-beijing' region.\n    at Request.ENOTFOUND_ERROR (/opt/outline/node_modules/aws-sdk/lib/event_listeners.js:507:46)\n    at Request.callListeners (/opt/outline/node_modules/aws-sdk/lib/sequential_executor.js:106:20)\n    at Request.emit (/opt/outline/node_modules/aws-sdk/lib/sequential_executor.js:78:10)\n    at Request.emit (/opt/outline/node_modules/aws-sdk/lib/request.js:688:14)\n    at ClientRequest.error (/opt/outline/node_modules/aws-sdk/lib/event_listeners.js:339:22)\n    at ClientRequest.<anonymous> (/opt/outline/node_modules/aws-sdk/lib/http/node.js:96:19)\n    at ClientRequest.emit (events.js:400:28)\n    at ClientRequest.emit (domain.js:475:12)\n    at TLSSocket.socketErrorListener (_http_client.js:475:9)\n    at TLSSocket.emit (events.js:400:28)\n    at TLSSocket.emit (domain.js:475:12)\n    at emitErrorNT (internal/streams/destroy.js:106:8)\n    at emitErrorCloseNT (internal/streams/destroy.js:74:3)\n    at processTicksAndRejections (internal/process/task_queues.js:82:21)","level":"error","message":"Error uploading to S3 from URL"}
{"error":"Inaccessible host: `inotea.s3.oss-cn-beijing.amazonaws.com'. This service may not be available in the `oss-cn-beijing' region.","stack":"UnknownEndpoint: Inaccessible host: `inotea.s3.oss-cn-beijing.amazonaws.com'. This service may not be available in the `oss-cn-beijing' region.\n    at Request.ENOTFOUND_ERROR (/opt/outline/node_modules/aws-sdk/lib/event_listeners.js:507:46)\n    at Request.callListeners (/opt/outline/node_modules/aws-sdk/lib/sequential_executor.js:106:20)\n    at Request.emit (/opt/outline/node_modules/aws-sdk/lib/sequential_executor.js:78:10)\n    at Request.emit (/opt/outline/node_modules/aws-sdk/lib/request.js:688:14)\n    at ClientRequest.error (/opt/outline/node_modules/aws-sdk/lib/event_listeners.js:339:22)\n    at ClientRequest.<anonymous> (/opt/outline/node_modules/aws-sdk/lib/http/node.js:96:19)\n    at ClientRequest.emit (events.js:400:28)\n    at ClientRequest.emit (domain.js:475:12)\n    at TLSSocket.socketErrorListener (_http_client.js:475:9)\n    at TLSSocket.emit (events.js:400:28)\n    at TLSSocket.emit (domain.js:475:12)\n    at emitErrorNT (internal/streams/destroy.js:106:8)\n    at emitErrorCloseNT (internal/streams/destroy.js:74:3)\n    at processTicksAndRejections (internal/process/task_queues.js:82:21)","level":"error","message":"Error uploading to S3 from URL"}
```

**Comment by frankli0324** [comment]
I also struggle to understand the logic behind these lines:

https://github.com/outline/outline/blob/e6eb43144cc9cdf1a505c95c118dec4f0d3742ab/server/utils/s3.ts#L25-L27

**Reply by tommoor** [reply]
Yes, it's the worst – open to improvements. It's trying to detect if the `AWS_S3_UPLOAD_BUCKET_URL` variable is a path-style url or not.

This configuration probably needs extracting so that it can be made testable with the wide variety of different configuration types that are out there.

**Reply by frankli0324** [reply]
how about just pass the configurations to the S3 constructor **as is**? could be breaking changes though, considering that it's a required configuration item
btw, the public endpoint for user access should also be configurable.

**Reply by tommoor** [reply]
> how about just pass the configurations to the S3 constructor as is? 

Yes, this would break many peoples existing installations

**Comment by frankli0324** [comment]
https://github.com/frankli0324/outline/blob/main/server/utils/s3.ts

Gave the backward compatibility up, but felt to be the right thing to do

**Reply by frankli0324** [reply]
May I open a PR and maybe add backward compatibility on top of this?

**Reply by tommoor** [reply]
Hi @frankli0324 yes, go for it – the level of thought you are putting in is very much appreciated and will help others in the future

**Comment by wohenbushuang** [comment]
I use Tencent cloud Cloud Object Storage (COS) as third-party S3 service, and all upload actions fail with 403 error. I replayed the request in Postman, and obtain the detail error message:

```xml
<?xml version='1.0' encoding='utf-8' ?>
<Error>
	<Code>AccessDenied</Code>
	<Message>Condition key x-cos-algorithm doesn&apos;t match the value AWS4-HMAC-SHA256</Message>
	<ServerTime>************</ServerTime>
	<Resource>BUCKET-NAME.cos.REGION.myqcloud.com/public/***************.png</Resource>
	<RequestId>*************************</RequestId>
	<TraceId>****************</TraceId>
</Error>
```

If I change the name of keys in the POST request body from `X-Amz-****` to `X-cos-****`, upload will succed.
```
X-Amz-Algorithm:AWS4-HMAC-SHA256
X-Amz-Credential:******************************/s3/aws4_request
X-Amz-Date:***********
X-Amz-Signature:**********
```

`X-Amz-Signature` and `X-cos-Signature` are both okay if one exists.

**Reply by tommoor** [reply]
There's probably a setting on COS side to accept the AMZ headers, otherwise it's not compatible

**Reply by wohenbushuang** [reply]
yep COS says they are "compatible" ... I have open a ticket with COS but I don't expect they will solve this. Just leave an experience in memory of my wasted time and for later visitors.

**Reply by frankli0324** [reply]
Unfortunately, alibaba OSS behaves similarly. I suspect that this is not part of `S3 Compatibility` for most (commercial?) object storage providers. I've also been asking the tencent COS team for compatibility improvements, but receives no answers.

**Reply by wohenbushuang** [reply]
Tencent COS gave me an example to show that their service is S3 compatible, and I have confirmed that it indeed works.
```js
import { S3 } from "aws-sdk";
const s3 = new S3({
  accessKeyId: 'xxx',
  secretAccessKey: 'xxx',
  endpoint: 'cos.ap-guangzhou.myqcloud.com',
});
s3.putObject({
  Bucket: 'test-xxx',
  Key: '1.txt',
  Body: Buffer.from('hello!'),
}, (err, data) => {
  if (err) {
    return console.error(err);
  }
  console.log('putObject:', data);
});
```
which will use standard `PUT` method.

I use docker to deploy outline `v0.62.0`, and I test attachment upload function by updating user avatar. I found that outline frontend will upload avatar with `POST` method, which is initiated by `app/util/uploadFile.ts:41` with parameters returned by `POST <my-outline-site>/api/attachments.create` (where calculated `X-Amz-***` comes from), instead of using standard `PUT` method in `aws-sdk` package. `s3.putObject` in `server/util/s3.ts` seems irrelavant.

![image](https://user-images.githubusercontent.com/15613184/161027873-3143e3af-a99e-47bb-b4bf-bbfe3ca1e98f.png)

https://github.com/outline/outline/blob/3ac125d560ece69d71c2d21e76ca1af2a0ac9112/app/utils/uploadFile.ts#L41-L44

I don't understand why should be so, and maybe this problem can get solved by directly using `aws-sdk` package dependence, or `PUT` method as [AWS S3 PutObject API](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html#API_PutObject_RequestSyntax) instructed. @frankli0324 @tommoor

**Reply by frankli0324** [reply]
outline uses `PostObject` API instead of `PutObject` for client file uploads, which I confirmed with tencent COS devs that are not compatible @wohenbushuang . The reason for users to upload files directly could be maximizing network performance and reduce server stress. The presigned POST request indeed comes from `aws-sdk`.
If you are suggesting that using a presigned `PutObject` URL, then outline wouldn't be able to control the file size. (At least from what I've learned...) ref to https://github.com/aws/aws-sdk-net/issues/424

https://github.com/outline/outline/blob/fea5f69a38d19244e70cac2e74e260cb84496a31/server/utils/s3.ts#L31

ref to:
https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-authentication-HTTPPOST.html 
https://cloud.tencent.com/document/product/436/14690

**Reply by wohenbushuang** [reply]
Well so sad to hear that. ~Seems we have to use minio as S3 gateway.~
Update: minio does not support COS either. https://blog.bosong.online/minio%E6%94%B9%E9%80%A0%E6%94%AF%E6%8C%81tencent-cos.html

**Reply by frankli0324** [reply]
感觉这就超出这个问题的范畴了
off-topic

**Reply by frankli0324** [reply]
COS will add `PostObject` compatibility in the next major release @wohenbushuang

**Comment by KooriMoe** [comment]
Well, here's a thing.
I remember that I tried to use BackBlaze, which has fully S3 protocol support, to set up as outline's storage services.
But like the Alibaba Cloud OSS, for avatar uploading, the outline can upload the images to the block, it cannot be set up as an avatar.
For the article's image uploading, it will have an error as I said in the summary of this discussion that the outline will change the bucket link to AWS S3 automatically, using the bucket name and endpoint to merge as an AWS service URL.

**Reply by almereyda** [reply]
You will need to set something like the `download` permission in Minio ¹, or like the "Block public access" turned off on AWS S3 ² with your provider.

¹ https://github.com/outline/outline/discussions/3281#discussioncomment-2430516
² https://wiki.generaloutline.com/share/125de1cc-9ff6-424b-8415-0d58c809a40f#h-permissions

**Reply by kerberjg** [reply]
@KooriMoe I managed to get in touch with Backblaze staff, here's what I've been told: https://github.com/outline/outline/discussions/2666#discussioncomment-3421528

To put it shortly, they currently don't support the type of uploads that Outline uses, but they might be implementing it in the future.
If you want to use Backblaze, I'd recommend you connect your B2 bucket to Minio Gateway proxy, and connect Outline to that
